pkg/controllers/chart/chart.go

```diff
@@ -226,11 +226,31 @@ func (c *Controller) OnRemove(key string, chart *v1.HelmChart) (*v1.HelmChart, e
 		return nil, nil
 	}
 
+	if chart.DeletionTimestamp == nil {
+		return nil, nil
+	}
+
 	expectedJob, objs, err := c.getJobAndRelatedResources(chart)
 	if err != nil {
 		return nil, err
 	}
 
+	// remove old helm-delete job if it was left
+	job, err := c.jobCache.Get(chart.Namespace, expectedJob.Name)
```

Comment:
I don't think you need to do this, the job patcher already deletes and re-creates the job as necessary, as part of the Apply logic:
https://github.com/k3s-io/helm-controller/blob/255f905f22ad5d303a387dae04abc7c3395283dd/pkg/controllers/chart/chart.go#L167-L173

Reply:
I guess the bug is due to `	WithOwner(chart).` in https://github.com/k3s-io/helm-controller/blob/255f905f22ad5d303a387dae04abc7c3395283dd/pkg/controllers/chart/chart.go#L242

First round of create & delete a HelmChart  X
```
HelmChart X with ID A 
  helm-delete job 1
  helm-delete job 2 (and possiblely left orphaned)
```

Second round of create & delete a HelmChart  X
```
Chart X with ID B
  helm-delete job 2 will not be replaced due to `WithOwner`
```
But https://github.com/k3s-io/helm-controller/blob/255f905f22ad5d303a387dae04abc7c3395283dd/pkg/controllers/chart/chart.go#L254 will be hit and  also https://github.com/k3s-io/helm-controller/blob/255f905f22ad5d303a387dae04abc7c3395283dd/pkg/controllers/chart/chart.go#L278 


That's the gap in the logic.

Reply:
If we manually create a same namespaced & named `helm-delete` job before deleting a HelmChart, I guess it will also shotcut the HelmChart's real deleting.

Reply:
If I remember correctly, the WithOwner stuff uses the owning object's name, namespace, and GVK to track resources across namespaces. As long as the owning chart shares those same attributes across delete/create cycles it should track it properly. If you run the controller with --debug you should get debug logs from the DesiredSet stuff showing you what it's trying to do - does that shed any light on the situation?

Reply:
@brandond 

I add some debug https://github.com/k3s-io/helm-controller/issues/177#issuecomment-2066423424 on Harvester environment, the logs /events can clearly show that, the `generic.ConfigureApplyForObject` is not able to correctly identify the old left `helm-delete` job. 

And the created `helm-delete` job, has no `Owner` field.

```
harv41:/home/rancher # kk get jobs -n cattle-logging-system helm-delete-rancher-logging -oyaml
apiVersion: batch/v1
kind: Job
metadata:
  annotations:
    batch.kubernetes.io/job-tracking: ""
    objectset.rio.cattle.io/applied: ...
    objectset.rio.cattle.io/id: helm-chart-registration
    objectset.rio.cattle.io/owner-gvk: helm.cattle.io/v1, Kind=HelmChart
    objectset.rio.cattle.io/owner-name: rancher-logging
    objectset.rio.cattle.io/owner-namespace: cattle-logging-system
  creationTimestamp: "2024-04-19T08:32:28Z"
  finalizers:
  - wrangler.cattle.io/promote-node-controller
  generation: 1
  labels:
    helmcharts.helm.cattle.io/chart: rancher-logging
    objectset.rio.cattle.io/hash: ea9e4622f05ef91896488127ad4be76f709ee325
  name: helm-delete-rancher-logging
  namespace: cattle-logging-system
  resourceVersion: "739101"
  uid: bcc55913-0ad5-48e9-8dcf-8b380f037ee8
spec:
  backoffLimit: 1000
  completionMode: NonIndexed
  completions: 1
  parallelism: 1
  selector:
    matchLabels:
      batch.kubernetes.io/controller-uid: bcc55913-0ad5-48e9-8dcf-8b380f037ee8
  suspend: false
  template:
    metadata:
      annotations:
        helmcharts.helm.cattle.io/configHash: SHA256=0832CE8EBAB899509A3AB82C345F3DF72C4D3300D5CDB2E0C47BB943F9630B34
      creationTimestamp: null
      labels:
        batch.kubernetes.io/controller-uid: bcc55913-0ad5-48e9-8dcf-8b380f037ee8
        batch.kubernetes.io/job-name: helm-delete-rancher-logging
        controller-uid: bcc55913-0ad5-48e9-8dcf-8b380f037ee8
        helmcharts.helm.cattle.io/chart: rancher-logging
        job-name: helm-delete-rancher-logging
    spec:
...  
  ```

Reply:
The orphaned job, was left per this return
```
	// once we have run the above logic, we can now check if the job is complete
	job, err = c.jobCache.Get(chart.Namespace, expectedJob.Name)
	if apierrors.IsNotFound(err) {
		// the above apply should have created it, something is wrong.
		// if you are here, there must be a bug in the code.
	@@ -269,7 +289,7 @@ func (c *Controller) OnRemove(key string, chart *v1.HelmChart) (*v1.HelmChart, e
		chartCopy.Status.JobName = job.Name
// the `chart` may be on on apiserver side
		newChart, err := c.helms.Update(chartCopy)
		if err != nil {
			return chart, fmt.Errorf("unable to update status of helm chart to add uninstall job name %s, %w", chartCopy.Status.JobName, err)
		}
```

And the normal running path has this call
```
	// note: an empty apply removes all resources owned by this chart
	err = generic.ConfigureApplyForObject(c.apply, chart, &generic.GeneratingHandlerOptions{
		AllowClusterScoped: true,
	}).
		WithOwner(chart).
		WithSetID("helm-chart-registration").
		ApplyObjects()
```

I will test if  `add this call to the above return` works.